mkosi/kmod.py

```diff
@@ -18,41 +20,101 @@
 def loaded_modules() -> list[str]:
     # Loaded modules are listed with underscores but the filenames might use dashes instead.
     return [
-        rf"/{line.split()[0].replace('_', '[_-]')}\.ko"
-        for line in Path("/proc/modules").read_text().splitlines()
+        normalize_module_name(line.split()[0]) for line in Path("/proc/modules").read_text().splitlines()
     ]
 
 
+def globs_match_filename(name, globs) -> Optional[bool]:
+    # Check whether the path matches any of the globs
+
+    for glob in reversed(globs):
+        # Patterns are evaluated in order and last matching one wins.
+        # Patterns may be prefixed with '-' to exclude modules.
+        if negative := glob.startswith("-"):
+            glob = glob[1:]
+        # As a special case, if a directory is specified, all items
+        # below that directory are matched.
+        if glob.endswith("/"):
+            glob += "*"
+
+        if (
+            # match the full path
+            (glob.startswith("/") and fnmatch.fnmatch(f"/{name}", glob))
+            or
+            # match a subset of the path, at path element boundary
+            ("/" in glob and fnmatch.fnmatch(f"/{name}", f"*/{glob}"))
+            or
+            # match the basename
+            fnmatch.fnmatch(name.split("/")[-1], glob)
+        ):
+            return not negative
+    return None
+
+
+def globs_match_module(name, globs) -> Optional[bool]:
+    # Strip '.ko' suffix and an optional compression suffix
+    name = re.sub(r"\.ko(\.(gz|xz|zst))?$", "", name)
+    # Check whether the suffixless-path matches any of the globs
+    return globs_match_filename(name, globs)
+
+
+def globs_match_firmware(name, globs) -> Optional[bool]:
+    # Strip any known compression suffixes
+    name = re.sub(r"\.(gz|xz|zst)$", "", name)
+    # Check whether the suffixless-path matches any of the globs
+    return globs_match_filename(name, globs)
+
+
 def filter_kernel_modules(
     root: Path,
     kver: str,
     *,
     include: Iterable[str],
     exclude: Iterable[str],
 ) -> list[Path]:
+    logging.debug(f"Kernel modules include: {' '.join(include)}")
+    logging.debug(f"Kernel modules exclude: {' '.join(exclude)}")
+
     modulesd = Path("usr/lib/modules") / kver
     with chdir(root):
-        modules = set(modulesd.rglob("*.ko*"))
+        # The glob may match additional paths.
+        # Narrow this down to *.ko, *.ko.gz, *.ko.xz, *.ko.zst.
+        modules = set(
+            m for m in modulesd.rglob("*.ko*") if m.name.endswith((".ko", ".ko.gz", ".ko.xz", ".ko.zst"))
+        )
```

Comment:
Nit: with the added logic this should be a set literal instead of a `set` call
```suggestion
        modules = {
            m for m in modulesd.rglob("*.ko*") if m.name.endswith((".ko", ".ko.gz", ".ko.xz", ".ko.zst"))
        }
```